crates/ty_python_semantic/src/types/constraints.rs

```diff
@@ -879,10 +867,9 @@ impl<'db> ConstrainedTypeVar<'db> {
 }
 
 #[derive(Clone, Debug, Eq, Hash, PartialEq, get_size2::GetSize, salsa::Update)]
-pub(crate) enum Constraint<'db> {
-    Range(RangeConstraint<'db>),
-    NotEquivalent(NotEquivalentConstraint<'db>),
-    Incomparable(IncomparableConstraint<'db>),
+pub(crate) struct Constraint<'db> {
+    positive: RangeConstraint<'db>,
+    negative: SmallVec<[NegatedRangeConstraint<'db>; 1]>,
```

Comment:
Is it worth (or have you already done) some experimentation on what performs best here? This optimizes for a single negated range -- is that the most common case? Is it better to use a regular `Vec` and optimize for zero negated ranges?

Reply:
I think that analysis won't be meaningful until #20093 lands (or as part of reviewing it). At the moment in real workloads we're still only creating constraint sets that are equivalent to `true` or `false`.

That said, my intuition is that this will be a common case, since there several negations in our various `has_relation_to` methods, and negating any range constraint would produce a constraint with one negated range in this field.